<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Question-1-:-The-$T=1$-case" data-toc-modified-id="Question-1-:-The-$T=1$-case-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Question 1 : The $T=1$ case</a></span><ul class="toc-item"><li><span><a href="#Question-1.1-:-Recognizing-a-known-problem" data-toc-modified-id="Question-1.1-:-Recognizing-a-known-problem-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Question 1.1 : Recognizing a known problem</a></span></li><li><span><a href="#Question-1.2-:-Solving-the-problem" data-toc-modified-id="Question-1.2-:-Solving-the-problem-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Question 1.2 : Solving the problem</a></span></li></ul></li><li><span><a href="#Question-2-The-$T=2$-case" data-toc-modified-id="Question-2-The-$T=2$-case-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Question 2 The $T=2$ case</a></span><ul class="toc-item"><li><span><a href="#Question-2.1" data-toc-modified-id="Question-2.1-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Question 2.1</a></span></li><li><span><a href="#Question-2.2-:-upper-bound" data-toc-modified-id="Question-2.2-:-upper-bound-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Question 2.2 : upper bound</a></span></li><li><span><a href="#Question-2.3-:-Splitted-variables-formulation" data-toc-modified-id="Question-2.3-:-Splitted-variables-formulation-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Question 2.3 : Splitted variables formulation</a></span></li><li><span><a href="#Question-2.4-:-anticipative-bound" data-toc-modified-id="Question-2.4-:-anticipative-bound-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Question 2.4 : anticipative bound</a></span></li><li><span><a href="#Question-2.5-:-two-stage-bound" data-toc-modified-id="Question-2.5-:-two-stage-bound-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Question 2.5 : two-stage bound</a></span></li></ul></li><li><span><a href="#Question-3-:-Heuristics-and-bounds-in-long-horizon" data-toc-modified-id="Question-3-:-Heuristics-and-bounds-in-long-horizon-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Question 3 : Heuristics and bounds in long horizon</a></span><ul class="toc-item"><li><span><a href="#Question-3.1" data-toc-modified-id="Question-3.1-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Question 3.1</a></span></li><li><span><a href="#Question-3.2-:-anticipative-bound" data-toc-modified-id="Question-3.2-:-anticipative-bound-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Question 3.2 : anticipative bound</a></span></li><li><span><a href="#Question-3.3-Writing-a-simulator" data-toc-modified-id="Question-3.3-Writing-a-simulator-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Question 3.3 Writing a simulator</a></span></li><li><span><a href="#Question-3.4" data-toc-modified-id="Question-3.4-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Question 3.4</a></span></li><li><span><a href="#Question-3.5" data-toc-modified-id="Question-3.5-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Question 3.5</a></span></li><li><span><a href="#Question-3.6" data-toc-modified-id="Question-3.6-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Question 3.6</a></span></li><li><span><a href="#Question-3.7-(Optional)" data-toc-modified-id="Question-3.7-(Optional)-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Question 3.7 (Optional)</a></span></li></ul></li><li><span><a href="#Question-4-Dynamic-Programming" data-toc-modified-id="Question-4-Dynamic-Programming-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Question 4 Dynamic Programming</a></span><ul class="toc-item"><li><span><a href="#Question-4.1-:-integrity-of-stock" data-toc-modified-id="Question-4.1-:-integrity-of-stock-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Question 4.1 : integrity of stock</a></span></li><li><span><a href="#Question-4.2" data-toc-modified-id="Question-4.2-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Question 4.2</a></span></li><li><span><a href="#Question-4.3" data-toc-modified-id="Question-4.3-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Question 4.3</a></span></li><li><span><a href="#Question-4.4" data-toc-modified-id="Question-4.4-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Question 4.4</a></span></li></ul></li></ul></div>

# The flower girl problem <a class="tocSkip">

Assume that every morning you can buy some flower at cost $c$, during the day you can sell them at price $p$. At the end of the day you can keep the flowers for the next day.
During each day the demand is an integer uniformly taken between $1$ and $D$.

At the start of your first day you have no flower in stock. At the end of the $T^{th}$ day all stock is lost. You want to minimize your expected loss summed over the $T$ days. We will denote $u_t$ the number of flower bought at the beginning of day $t$. We assume that $u_t\in [0,U]$ (in particular we assume that it is continuous).

The parameters are set as follow.

In [1]:
# If not installed yet
import Pkg
Pkg.add("JuMP")
Pkg.add("GLPK")

    Updating registry at `C:\Users\jc\.julia\registries\General`
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Resolving package versions...
  No Changes to `C:\Users\jc\.julia\environments\v1.6\Project.toml`
  No Changes to `C:\Users\jc\.julia\environments\v1.6\Manifest.toml`
   Resolving package versions...
  No Changes to `C:\Users\jc\.julia\environments\v1.6\Project.toml`
  No Changes to `C:\Users\jc\.julia\environments\v1.6\Manifest.toml`


In [2]:
using JuMP, GLPK, Statistics

OPTIMIZER = GLPK.Optimizer

# problem parameters
c = 1  # cost of buying
p = 2  # cost of selling
D = 10 # demand is uniform over 1:D
U = 7  # maximum buy per day

7

## Question 1 : The $T=1$ case

### Question 1.1 : Recognizing a known problem

If $T=1$, recognize the flower girl problem as a known optimization problem ? 

Newsvender!

### Question 1.2 : Solving the problem

Write the extensive formulation of the problem as an LP and solve it using `JuMP` and a linear solver.

##### Solution 

The extensive formultaion of the problem is:- 
$$Min ~~~ P\sum_{i = 1}^{D} cu - pv_i$$
Subject to
$$u \leq U$$
$$v_i \leq u$$
$$v_i \leq d_{Ti}$$

Where,
<ol>
<li> $P = \frac{1}{D}~$ </li>
<li> $u \equiv $ Quantity of flowers to be bought at the beginning of the day </li>
<li> $v_i \equiv $ Quantity of sold flowers at the end of the day for each scenario $i$ </li>
<ol>

In [15]:
T = 1
P = 1/D 

d = zeros(T,D)
for s = 1:D
    d[:,s] .= s # d[1,s] is the demand at time 1 for scenario s
end

In [16]:
# lINEAR PROGRAM 

# Defining the MODEL
m = Model(with_optimizer(OPTIMIZER))  

# Variables 
@variable(m, u>=0)                      # The number of the bought flowers must be non-negative 
@variable(m, v[1:D]>=0)                 # The number of sold flowers couldn't be negative 

# Objective Function
@objective(m,Min, P*sum(c*u-p*v[i] for i=1:D))             # Minimizing the COST function 

# Constraints 
@constraints(m,begin 
            u <= U                                 # We can't buy more than 7 flowers 
            [i in 1:D], v[i] <= u                  # We can't sell more than what we have already bought 
            [i in 1:D], v[i] <= d[T,i]             # We can't sell more than the DEMAND 
    end
    )

optimize!(m)

In [17]:
# RESULTS 

println("Objective Value = ",objective_value(m))
println("u = ",value.(u))
print("v = ",value.(v))

Objective Value = -3.0000000000000018
u = 5.0
v = [1.0, 2.0, 3.0, 4.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0]

## Question 2 The $T=2$ case

We consider now that $T=2$.

### Question 2.1 
From Question 1 deduce an upper bound to this problem.

###### Solution 

The **upper bound** is indeed the **objective value** of $Q1.2$ (<font color='red'>$UP = -3.0000000000000018 \approx -3$</font>)

Because, in this case if you didn't sell all the flowers in the $1^{st}$ day, you are able to keep them for the $2^{nd}$ day. Hence there is a probability of selling more flowers and therefore gain more money (minimizing your cost) even if you didn't buy any flowers at the begining of the $2^{nd}$ day. 

Thus, **at least** you will gain the same as the $1^{st}$ day.

### Question 2.2 : upper bound
Solve the flower-girl problem with a stochastic programming approach, by solving the associated extensive formulation.

##### Solution 

The extensive formultaion of the problem is:- 
$$Min ~~~ P\sum_{i = 1}^{D} cu - pz_i + P^2\sum_{i = 1}^{D} \sum_{j = 1}^{D} cw_i - pv_{ij}$$
Subject to
$$u \leq U ~~~~ \dots\dots (1)$$ 
$\forall ~i = 1 \dots D$
\begin{cases} w_i \leq U ~~~~ \dots\dots (2)\\ 
z_i \leq u ~~~~ \dots\dots (3)\\
w_i \leq d_{1i} ~~~~ \dots\dots (4)\\ 
z_i \leq d_{1i} ~~~~ \dots\dots (5)
\end{cases}
$\forall ~i,j = 1 \dots D$
\begin{cases}
v_{ij} \leq d_{2i} ~~~~ \dots\dots (6)\\ 
v_{ij} \leq u - z_i + w_i ~~~~ \dots\dots (7)
\end{cases}



Where,
1. $P = \frac{1}{D}~$ </li>
2. $u \equiv $ Quantity of flowers to be bought at the beginning of the $1^{st}$ day 
3. $z_i \equiv$ Quantity of sold flowers at the end of the $1^{st}$ day for each scenario $i$
3. $w_i \equiv $ Quantity of flowers to be bought at the beginning of the $2^{nd}$ day for each scenario $i$
4. $v_{ij} \equiv$ Quantity of sold flowers at the end of the $2^{nd}$ day for each scenario $i$ and each scenario $j$

**Interpretation of the Constraints**
1. Constraints (1) & (2): We can't buy more that 7 flowers in any day.
2. Constraint (3): At the $1^{st}$ day, we can't sell more than what we have already bought at the beginning of the day.
3. Constraints (4) & (5): $z_i ~\& ~ w_i$ are measurable w.r.t $d_1$.
4. Constraints (5) & (6): We can't sell more than the demand in any day.
5. Constraint (6): $v_{ij}$ is measurable w.r.t $d_1$ as well as $d_2$.
6. Constraint (7): At the $2^{nd}$ day, we can't sell more than what we have bought at the beginning of the $2^{nd}$ day in addition to the remaining of the $1^{st}$ day ($i.e$ the difference between how much we bought and how much we sold).

In [18]:
T = 2
P = 1/D
PP = 1/D^2

d = zeros(T,D) #d(t,s) is the demand for realisation s at time t
for s in 1:D
    d[:,s] .= s 
end

In [20]:
# LINEAR PROGRAM

# Defining the MODEL 
m = Model(with_optimizer(OPTIMIZER))

# Variables 
@variable(m,u>=0)                   # Number of flowers to be bought at the beginning of the first day                   
@variable(m,w[1:D]>= 0)             # Number of flowers to be bought at the beginning of the second day         
@variable(m,z[1:D]>= 0)             # Number of the sold flowers at the end of the first day         
@variable(m,v[1:D, 1:D]>=0)         # Number of the sold flowers at the end of the second day

# Objective Function 
@objective(m,Min, (P*sum(c*u - p*z[i] for i=1:D)) + (PP*sum(c*w[i] - p*v[i,j] for i=1:D , j=1:D)))  # Minimizing the COST function    

# Constraints 
@constraint(m, u <= U)

for i=1:D
    @constraint(m, w[i] <= U)
    @constraint(m, z[i] <= d[1,i])
    @constraint(m, z[i] <= u)
    @constraint(m, w[i] <= d[1,i])
end

for i=1:D , j=1:D
    @constraint(m, v[i,j] <= d[2,j])
    @constraint(m, v[i,j] <= u - z[i] + w[i])
end

optimize!(m)

In [27]:
# RESULTS 

println("Objective Value = ",objective_value(m))
println("u = ",value.(u))
println("z = ",value.(z))
println("w = ",value.(w))
println("v = ")
show(stdout, "text/plain", value.(v))

Objective Value = -7.899999999999992
u = 7.0
z = [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 7.0, 7.0, 7.0]
w = [0.0, 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 5.0, 5.0, 5.0]
v = 
10×10 Matrix{Float64}:
 1.0  2.0  3.0  4.0  5.0  6.0  6.0  6.0  6.0  6.0
 1.0  2.0  3.0  4.0  5.0  5.0  5.0  5.0  5.0  5.0
 1.0  2.0  3.0  4.0  5.0  5.0  5.0  5.0  5.0  5.0
 1.0  2.0  3.0  4.0  5.0  5.0  5.0  5.0  5.0  5.0
 1.0  2.0  3.0  4.0  5.0  5.0  5.0  5.0  5.0  5.0
 1.0  2.0  3.0  4.0  5.0  5.0  5.0  5.0  5.0  5.0
 1.0  2.0  3.0  4.0  5.0  5.0  5.0  5.0  5.0  5.0
 1.0  2.0  3.0  4.0  5.0  5.0  5.0  5.0  5.0  5.0
 1.0  2.0  3.0  4.0  5.0  5.0  5.0  5.0  5.0  5.0
 1.0  2.0  3.0  4.0  5.0  5.0  5.0  5.0  5.0  5.0

### Question 2.3 : Splitted variables formulation

We give the flower-girl problem for $2$ time step with splitted variables.
Check that you obtain the same value as before.  

In [7]:
T = 2

d = zeros(T,D,D) # d(t,s,ss) is the demand for scenario (s,ss) at time t
for s = 1:D, ss = 1:D
    d[1,s,ss] = s
    d[2,s,ss] = ss
end


m = Model(with_optimizer(OPTIMIZER))

@variables(m,begin
    0<= u[1:T,1:D,1:D] <= U  # bought flowers at beginning of day t
    v[1:T,1:D,1:D] >= 0      # sold flowers during day t
    end)     

@objective(m, Min, 1/D^2*sum(c*u[t,s,ss]-p*v[t,s,ss] for t = 1:T, s = 1:D, ss = 1:D))

for t = 1:T, s = 1:D, ss = 1:D
    @constraint(m, v[t,s,ss] <= d[t,s,ss])                     # we sell no more than demand
    @constraint(m, sum(v[tt,s,ss]-u[tt,s,ss] for tt=1:t) <= 0) # we sell no more than stock
end

    ### u1 is deterministic - i.e. the same for all scenarios
for s = 1:D, ss = 1:D
    @constraint(m,u[1,s,ss] == 1/D^2 *sum( u[1,t,tt] for t=1:D, tt=1:D))
end
    ### u2 and v1 are d1 measurable - i.e. does not differentiate on ss
for s = 1:D, ss = 1:D
    @constraint(m,u[2,s,ss] == 1/D *sum( u[2,s,tt] for tt=1:D))
    @constraint(m,v[1,s,ss] == 1/D *sum( v[1,s,tt] for tt=1:D))
end
       
optimize!(m)
@show(objective_value(m))
#round.(value.(u[1,:,:]),digits=2)
                       

objective_value(m) = -7.899999999999949


-7.899999999999949

### Question 2.4 : anticipative bound

By suppressing constraints in the splitted version, find the lower-bound given by the anticipative approach.


###### Solution 

We should relax all the measurability constraints in the above code 

$i.e.~$ I am going to assume that $u_1, u_2 ~\&~ v_1$ are measurable w.r.t. both $d_1 ~\&~ d_2$

Thus, deleting all the follwing constraints:-
1. @constraint(m,u[1,s,ss] == 1/D^2 *sum( u[1,t,tt] for t=1:D, tt=1:D))
2. @constraint(m,u[2,s,ss] == 1/D *sum( u[2,s,tt] for tt=1:D))
3. @constraint(m,v[1,s,ss] == 1/D *sum( v[1,s,tt] for tt=1:D))

In [24]:
T = 2

d = zeros(T,D,D)   # d(t,s,ss) is the demand for scenario (s,ss)
for s = 1:D, ss = 1:D
    d[1,s,ss] = s
    d[2,s,ss] = ss
end


m = Model(with_optimizer(OPTIMIZER))

@variables(m,begin
    0<= u[1:T,1:D,1:D] <= U    # bought flowers at beginning of day t
    v[1:T,1:D,1:D] >= 0        # sold flowers during day t
    end)     

@objective(m, Min, 1/D^2*sum(c*u[t,s,ss]-p*v[t,s,ss] for t = 1:T, s = 1:D, ss = 1:D))

for t = 1:T, s = 1:D, ss = 1:D
    @constraint(m, v[t,s,ss] <= d[t,s,ss])                        # we sell no more than demand
    @constraint(m, sum(v[tt,s,ss]-u[tt,s,ss] for tt=1:t) <= 0)    # we sell no more than stock
end
       
optimize!(m)
@show(objective_value(m))
round.(value.(u[1,:,:]),digits=2)

objective_value(m) = -10.11999999999997


10×10 Matrix{Float64}:
 1.0  1.0  1.0  1.0  1.0  1.0  1.0  2.0  3.0  4.0
 2.0  2.0  2.0  2.0  2.0  2.0  2.0  3.0  4.0  5.0
 3.0  3.0  3.0  3.0  3.0  3.0  3.0  4.0  5.0  6.0
 4.0  4.0  4.0  4.0  4.0  4.0  4.0  5.0  6.0  7.0
 5.0  5.0  5.0  5.0  5.0  5.0  5.0  6.0  7.0  7.0
 6.0  6.0  6.0  6.0  6.0  6.0  6.0  7.0  7.0  7.0
 7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0
 7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0
 7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0
 7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0  7.0

### Question 2.5 : two-stage bound

By suppressing constraints in the splitted version, find the lower-bound given by the 2 stage-approach.

###### Solution 

In the **two-stage** structure, we still have that $u_1$ is deterministic but not the other variables. 

$i.e.~$ I am going to assume that $u_2 ~\&~ v_1$ are measurable w.r.t. both $d_1 ~\&~ d_2$

Thus, deleting the follwing constraints:-
1. @constraint(m,u[2,s,ss] == 1/D *sum( u[2,s,tt] for tt=1:D))
2. @constraint(m,v[1,s,ss] == 1/D *sum( v[1,s,tt] for tt=1:D))

In [28]:
T = 2

d = zeros(T,D,D) # d(t,s,ss) is the demand for scenario (s,ss)
for s = 1:D, ss = 1:D
    d[1,s,ss] = s
    d[2,s,ss] = ss
end


m = Model(with_optimizer(OPTIMIZER))

@variables(m,begin
    0<= u[1:T,1:D,1:D] <= U  # bought flowers at beginning of day t
    v[1:T,1:D,1:D] >= 0      # sold flowers during day t
    end)     

@objective(m, Min, 1/D^2*sum(c*u[t,s,ss]-p*v[t,s,ss] for t = 1:T, s = 1:D, ss = 1:D))

for t = 1:T, s = 1:D, ss = 1:D
    @constraint(m, v[t,s,ss] <= d[t,s,ss])                     # we sell no more than demand
    @constraint(m, sum(v[tt,s,ss]-u[tt,s,ss] for tt=1:t) <= 0) # we sell no more than stock
end

    ### u1 is deterministic - i.e. the same for all scenarios
for s = 1:D, ss = 1:D
    @constraint(m,u[1,s,ss] == 1/D^2 *sum( u[1,t,tt] for t=1:D, tt=1:D))
end

optimize!(m)
@show(objective_value(m))
round.(value.(u[2,:,:]),digits=2)

objective_value(m) = -9.769999999999959


10×10 Matrix{Float64}:
 0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  3.0  4.0
 0.0  0.0  0.0  0.0  0.0  1.0  2.0  3.0  4.0  5.0
 0.0  0.0  0.0  0.0  1.0  2.0  3.0  4.0  5.0  6.0
 0.0  0.0  0.0  1.0  2.0  3.0  4.0  5.0  6.0  7.0
 0.0  0.0  1.0  2.0  3.0  4.0  5.0  6.0  7.0  7.0
 0.0  1.0  2.0  3.0  4.0  5.0  6.0  7.0  7.0  7.0
 1.0  2.0  3.0  4.0  5.0  6.0  7.0  7.0  7.0  7.0
 1.0  2.0  3.0  4.0  5.0  6.0  7.0  7.0  7.0  7.0
 1.0  2.0  3.0  4.0  5.0  6.0  7.0  7.0  7.0  7.0
 1.0  2.0  3.0  4.0  5.0  6.0  7.0  7.0  7.0  7.0

## Question 3 : Heuristics and bounds in long horizon

We assume now that we work on a week, that is $T=10$. 

### Question 3.1  
how many variables have the extended formulation ? 

###### Solution 
It's exactly $12222222221$ variables.

**Explanation**
For $T = 2$ 
we have:-  (I used the same notations as in $Q2$)
1. 1 variable at the beginning of the $1^{st}$ day  ($u$)
2. 10 variables at the end of the $1^{st}$ day   ($w_i$)
3. 10 variables at the beginning of the $2^{nd}$ day   ($z_i$)
4. 100 variables at the end of the $2^{nd}$ day  ($v_{ij}$)

Thus, a sum equal to 121 

Repeating the same process for $T = 10$, we have the result.

**Using the approximation results in the lecture notes**
1. $T = 10$
2. $n_d = 10$ (10 scenarios for each day)
3. $n_u = 2$  ($U_t(x) = \{u,v\}$ s.t. $u \equiv$ how much we bought at the beginning of the day $t$ & $v \equiv$ how much we sold at the end of the day $t$)

Thus, # variables $\approx ~ 2(10^{10})$ 

### Question 3.2 : anticipative bound

Estimate (by Monte Carlo) the lower-bound given by the anticipative approach.

In [29]:
T = 10
Nmc = 10^3

## We define an helper function that print a Monte Carlo estimate from a vector of results
function mc(res, print_opt=true)
    Nmc = length(res)
    print_opt && print(round(mean(res);digits=2)," +/- ",round(1.96*std(res)/sqrt(Nmc);digits=2)) 
    return round(mean(res);digits=2)
end

### scenarios of Monte-Carlo test for fair comparison
using Random
Random.seed!(0)
scenarios_mc = rand(1:D,(Nmc,T)); # generate Nmc scenarios of demand

###### Solution 

I'm going to formulate the problem as a linear program and solve $Nmc = 1000$ deterministic problems by using the extended formulation

For each scenario $j = 1 \dots \dots Nmc$, the problem reads 
$$Min ~ \sum_{i = 1}^{T} cu_i - pv_i$$

Subject to 

$\forall ~ i = 1 \dots\dots T$
\begin{cases}
u_i \leq U \\
v_i \leq \text{scenarios}_{ij} 
\end{cases}
$$v_1 \leq u_1$$
$$v_i \leq u_i + \sum_{s = 1}^{i-1} u_s - v_s ~~~~~ \forall i = 2 \dots T$$

Where, 

1. $u_i \equiv $ Quantity of flowers to be bought at the beginning of the $i^{th}$ day 
2. $v_i \equiv$ Quantity of sold flowers at the end of the $i^{th}$ day

In [36]:
function anticipative(scenarios)
    #compute the anticipative bound obtained from scenarios in d
    res = zeros(Nmc,1) # store the value of each scenario

    for j=1:Nmc
        
        # Defining the MODEL
        m = Model(with_optimizer(OPTIMIZER))
        
        # Variables 
        @variable(m,u[1:T] >= 0)     # Quantity of flowers to be bought at the beginning of the 𝑖𝑡ℎ day
        @variable(m,v[1:T] >= 0)     # Quantity of sold flowers at the end of the  𝑖𝑡ℎ  day
        
        # Objective Function 
        @objective(m,Min,sum(c*u[i] - p*v[i] for i=1:T))   # Minimizing the COST function 
        
        # Constraints 
        @constraint(m,v[1] <= u[1])                     # At the 1st day, we can't sell more than what we've already bought 
        for i=1:T
            @constraint(m, u[i] <= U)                   # At each day, we can't buy more than 7 flowers 
            @constraint(m, v[i] <= scenarios[j,i])      # At each day, we can't sell more than the demand 
        end
        
        # From the 2nd day, we can't sell more than what we bought at the beginning of the day in addition to 
        #                     the remaining flowers from the past days 
        for i=2:T
            @constraint(m, v[i] <= u[i] + sum(u[s]-v[s] for s=1:i-1))  
        end

        optimize!(m)
        res[j] = objective_value(m)     # Inserting the objective value for each scenario j 
    end            

    return mc(res)
end

O = anticipative(scenarios_mc)
println()
print("Objective value = ",O)

-53.27 +/- 0.52
Objective value = -53.27

### Question 3.3 Writing a simulator

Write a simulator that evaluate a policy over a given set of scenarios.

A policy is a function taking as argument $(t,x)$, $t$ being the day, $x$ the amount of newspaper in stock at the beginning of day $t$ and return $u$ the quantity of newspaper bought in the morning of day $t$. The quantity of newspaper sold during day $t$ will be computed as $\min(x+u,d)$. 



###### Solution 

I'm going to compute the optimal cost for each scenario ($Nmc = 1000$ scenarios) and then return the value of the empirical cost by using monte carlo 

For each scenario $i = 1 \dots \dots Nmc$ 

$$costs ~=~ \sum_{j = 1}^{T} cu_j - p \times min(S_j + u_j, scenarios_{ij})$$

Where, 

1. The control $u_j$ is to be determined by using a specific policy 
2. The stock $S_j$ is equal to 0 at the beginning of the $1^{st}$ day ($i.e. ~ S_1 = 0$), then from the $2^{nd}$ day it will be updated using the following formula $$S_{j + 1} = s_j + u_j - min(S_j + u_j, scenarios_{ij}) ~~~ \forall ~ j = 1 \dots (T-1)$$

In [67]:
function simulator(scenarios,policy)
    costs = zeros(Nmc) # vector of total cost for each of the Nmc scenarios    
    u = zeros(T)
    
    for i=1:Nmc               
        S = 0                      # The stock (S) at the beginning of the 1st day is equal to 0 
        for j=1:T
            u[j] = policy(j, S)    # Determining the control u for each day j by using a specific policy 
            costs[i] = costs[i] + (c*u[j] - p*min(S + u[j], scenarios[i,j]))    # Computing the cost for each day j 
            S = S + u[j] - min(S + u[j], scenarios[i,j])                        # updating the stock for the next day 
            S = Int(round(S))
        end
    end
            
    return mc(costs,false) # return the mean and print the estimation of costs
end

#simulator(scenarios_mc,control_2S)

simulator (generic function with 1 method)

### Question 3.4

Evaluate the heuristic consisting in buying the same amount $u$ of flower everyday, for $u = 0..m$, and give an upper bound to the problem.

###### Solution 

For any value of the control $u$ your input, the defined policy (same_u) will just return your input.

Then, using the **simulator** we got the cost. 

Indeed, this policy is an upper bound to the problem

In [74]:
m = readline()
m = parse(Float64, m)

function same_u(t, x)    
    return m
end

stdin> 6


same_u (generic function with 1 method)

An upper bound:- 

In [75]:
O = simulator(scenarios_mc,same_u)
print("Objective Value = ", O)

Objective Value = -41.86

### Question 3.5

Evaluate (by Monte Carlo) a model predictive approach of the problem using the expectation of demand as prediction. Compare it with using the best constant $u$.

###### Solution 

First of all, recall that the mean of the Uniform Distribution $U\{a,b\}$ is $\frac{a+b}{2}$

Thus, the expectation of the demand which has a uniform distribution $U\{1,D\}$ is $E = \frac{1 + D}{2}$

Now, using the MPC, the demand starting from $t$ is equal to $$d = E \times (T - t + 1)$$

Thus, the MPC policy works as follows 

$$MPC(t,stock) =
\begin{cases}
d - stock  & \text{if} ~~ d - stock > U \\
U & \text{Otherwise} 
\end{cases} $$

In [46]:
## Define an MPC policy using the mean
function MPC_control(t,stock)
    # return the value of the MPC control at time t starting from stock
    E = (1 + D)/2
    d = E*(T - t + 1)
    return min(U, d - stock)
end

MPC_control (generic function with 1 method)

In [47]:
simulator(scenarios_mc,MPC_control)

-50.87

In [76]:
# For the aim of comparison between the same_u and the MPC ploices with control value = 5.5 in the same_u policy 
m = 5.5 
function same_u(t, x)    
    return m
end
simulator(scenarios_mc,same_u)

-42.2

###### Comparison 
MPC is more efficient than choosing the same control which is logical since it has more flexibility of choosing $u$ for each $t$ and for each scenario.

### Question 3.6 

Give an estimated lower bound obtained from a two-stage approximation. 

Two-stage approximation can be implemented by the following linear program 

$$Min ~~ P\sum_{i = 1}^{Nmc} \sum_{j = 1}^{T} cu_{ij} - pv_{ij}$$

Subject to 

$\forall i = 1 \dots \dots Nmc, \forall j = 1 \dots \dots T$
\begin{cases}
u_{ij} \leq U ~~~~~~ \dots \dots (1) \\
v_{ij} \leq scenarios_{ij} ~~~~~~ \dots \dots (2)
\end{cases}
$\forall i = 1 \dots \dots Nmc$
\begin{cases}
u_{i1} = P\sum_{s = 1}^{Nmc} u_{s1}  ~~~~~~ \dots \dots (3) \\ 
v_{i1} \leq u_{i1}        ~~~~~~ \dots \dots (4)
\end{cases}
$\forall i 1 \dots \dots Nmc, j = 1 \dots \dots (T-1)$
$$v_{i,j+1} \leq u_{i, j+1} + \sum_{s = 1}^{j} u_{is} - v_{is}  ~~~~~~ \dots \dots (5)$$


Where,
1. $P = \frac{1}{Nmc}$ 
2. $u_{ij} \equiv $ Quantity of flowers to be bought at the beginning of the $i^{th}$ day for the $j^{th}$ scenario  
3. $v_{ij} \equiv$ Quantity of sold flowers at the end of the $i^{th}$ day for the $j^{th}$ scenario 

**Interpretation of the Constraints**
1. Constraints (1): We can't buy more that U flowers in any day and in any scenario.
2. Constraint (2): We can't sell more than the demand in any day and in any scenario.
3. Constraints (3): $u_{i1}$ the contorl of the $1^{st}$ day (how much to bought at the beginning of the $1^{st}$ day) is **deterministic** for each scenario $i$
4. Constraints (4): At the $1^{st}$ day, we can't sell more than what we have bought at the beginning of the $1^{st}$ day.
6. Constraint (5): For each scenario $i$, and from the 2nd day, we can't sell more than what we bought at the beginning of the day in addition to the remaining flowers from the past days. 

In [58]:
function two_stage(scenarios)
    
    P = 1/Nmc
    
    # Defining the MODEL 
    m = Model(with_optimizer(OPTIMIZER))
    
    # Variables 
    @variable(m, u[1:Nmc,1:T] >= 0)
    @variable(m, v[1:Nmc,1:T] >= 0)
    
    # Objective Function 
    @objective(m, Min, P*sum(c*u[i,j] - p*v[i,j] for i=1:Nmc, j=1:T))   # Minimizing the COST function 

    # Constraints 
        for i=1:Nmc, j=1:T
            @constraint(m, u[i,j] <= U)                    # We can't buy more than U flowers for any (i,j)
            @constraint(m, v[i,j] <= scenarios[i,j])      # We can't sell more than the demand for any (i,j)  
        end
        
        for i=1:Nmc
            @constraint(m, u[i,1] == P*sum(u[s,1] for s=1:Nmc))   # u1 is deterministic for each scenario i
            @constraint(m, v[i,1] <= u[i,1])                      # At the 1st day, we can't sell more than what we have bought 
        end
    
    # For each scenario 𝑖, and from the 2nd day, we can't sell more than what we bought at the beginning of the day
    #                               in addition to the remaining flowers from the past days.
        for i=1:Nmc, j=1:(T-1)
            @constraint(m, v[i,j+1] <= u[i,j+1] + sum(u[i,s] - v[i,s] for s=1:j))
        end
    
    optimize!(m)
    return objective_value(m)
          
end

two_stage (generic function with 1 method)

In [59]:
two_stage(scenarios_mc)

-53.27400000000202

### Question 3.7 (Optional) 

Evaluate the repeated two-stage approach (start with a small amount of scenarios).

In [ ]:
#YOUR CODE HERE
error("to be implemented")

## Question 4 Dynamic Programming


### Question 4.1 : integrity of stock
Show that the optimal stock at the end of day is integer and no more than $D$.

I tried solving this question. Unfortunately, I didn't get a complete proof of it 

However, I'll anyway write what I did

1. For me I think we can use the following argument to prove it 
"An optimal state trajectory is obtained by $x_{t+1} = \psi_t^V(\mathbf{x_t})$, with"
$$\psi_t^V(x,d) \in argmin_{y \in \mathbf{X_t}(x,d)} ~~ c_t(x,y,d) + V_{t+1}(y)$$
2. Finding the values of \Psi that minimizes the above argument implies the optimal values of $x$ since $x_{t+1} = \psi_t^V(\mathbf{x_t})$
3. To find the minimizers of the above argument we should find the derivative w.r.t $y$ and calculate the values of $y$ with the derivative equals 0 

$i.e.$ $$\nabla_y c_t(x,y,d) + \nabla_y V_{t+1}(y) = 0 ~~~~ \forall t$$

4. I came up with a formula for the function $c$ but I wasn't able for $V$ 

For $c$, 

$$c(x,y,d) = c(y - x + d) - p \times min(y + d, d) $$
$$~~~~~~~~~~~~~~~~~~~~~~~~ = c(y - x + d) - p \begin{cases} y + d & \text{If} ~ y < 0 \\ d & \text{Otherwise} \end{cases}$$
$$~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ = c(y - x + d) - p \begin{cases} (c - p)y - x + (1 - p)d & \text{If} ~ y < 0 \\ cy - x + (1 - p)d & \text{Otherwise} \end{cases}$$


Thus, $$\nabla_y c(x,y,d) = \begin{cases} c - p & \text{If} ~ y < 0 \\ c & \text{Otherwise} \end{cases}$$

### Question 4.2 
How many operations are required to solve the problem by Dynamic Programming ? How many scenarios for an extended formulation ?

1. $T = 10$ 
2. From $Q4.1$, the state $~x \in \mathbb{X} = \{0, 1, 2, \dots \dots, D\} \Rightarrow |\mathbb{X}| = 11$
3. The controls $~ u \in [0,U]$. However, since again by $Q4.1,~ x$ takes just an integer values so indeed the control $u$ takes an integer values as well. Thus $u \in \mathbb{U} = \{0, 1, 2, \dots, U\} \Rightarrow |\mathbb{U}| = 8$
4. The demand $d \in \mathbb{D} = \{1, 2, \dots \dots, D\} \Rightarrow |\mathbb{D}| = 10$

**Thus,** the number of operations is 
$$\mathcal{O}(10 \times 11 \times 8 \times 10)$$

* According the number of **scenarios** for an extended formulation, it's $10 \times 11 \times 8 \times 10 = 8800$ sceanrios)

### Question 4.3
Find the optimal value of the flower-girl problem by Dynamic Programming. Describe the optimal strategy.

###### Solution 

I implemented ,almost, algorithm 3 (Classical stochastic dynamic programming algorithm) presented in the lecture notes 

By using Bellman's equations
$\begin{cases}
V_T(x) = K(x) \\
\hat{V}_t(x,d) = min_{y \in \mathbf{X_t}(x,d)} c_t(x, y, d) + V_{t+1}(y) \\ 
V_t(x) = \mathbb{E}[\hat{V}_t(x, d_t)]
\end{cases}$

With:- 
$$c_t(x, y, d) = c_t(x, u, d) = cu - p \times min(x + u, d)$$ 
Where, $y = x + u - d$ such that x: is the stock, u: is the control (How much flowers we will bought at the beginning of the day)

**Now**, I defined an $(T + 1) \times (D + 1)$ zero matrix $V$ (matrix of costs) such that:- 
1. For all $t = 1 \dots \dots T$, and for any $x$, $v_{tx}$ is the cost being at the $t^{th}$ day with stock $x$ at the beginning of the day.
2. For $t = T + 1$ and for any $x$, $v_{T+1,x}$ is the cost being at the end of the day $T$ with stock $x$. Thus, $V_{T+1,x} = 0 = K(x), ~~ \forall x$

**Thus,** Bellman's equations read:- 
$\begin{cases}
V_{T+1}(x) = 0 \\
\hat{V}_t(x,d) = min_{u} ~~ [cu - p \times min(x + u, d)] + V_{t+1}(x + u - d) \\ 
V_t(x) = P\sum_{d = 1}^{D} \hat{V}_t(x, d)
\end{cases}$

Where, $P = \frac{1}{D}$

**Finally,** I have used the result from $Q4.1$, so that any stock with value $> D$ has a cost goes to $\infty$

In [64]:
V = zeros(T+1,D+1)    # Defining the matrix of costs with zeros, to be updated 
P = 1/D

#(for t=1:T is indeed a "Backward loop" because of indexing the V as T-t+1, so if t = 1, then we compute V[T-1+1] = V[T])
#                    the smae for t = T, V[T - T + 1] = V[1]
for t=1:T                       # Loop of days  
    for x=0:D                   # Loop of stocks 
        V_hat = Inf             # In order to minimize 
        for u=0:U      # Loop of controls (It's before the loop of demands, since we buy the flowers before knowing the demand)
            z = 0
            for d=1:D              # Loop of demands 
                y = x + u - d      # The new state (i.e. the stock for the next day)
            
# If the stock has a value between 1 and D, it yields for an optimal cost (Q4.1's result), so we compute the cost
                if (y>0 && y<=D)        
                    v = (c*u - p*min(x+u,d)) + V[T-t+2,y+1]
# The stock can't be negative, so if y is non-positive, the cost computed by the same equation with V[.,1] (cost when x = 0) 
                elseif y <= 0
                    v = (c*u - p*min(x+u,d)) + V[T-t+2,1]
# If y > D, then we are sure that it isn't the optimal solution, so we can assume that the cost goes to infinty 
                else
                    v = Inf
                end
                z += v             # Summing the costs for all d 
            end 
            if z < V_hat          # Choosing the best (minimum) cost for all the controls u 
                V_hat = z
            end
        end
        V[T-t+1,x+1] = P*V_hat    # Updating the value function 
    end
end

println(round.(V[1,:], digits=3))      #Printing the costs for the 1st day for any stock x 

[-50.663, -52.063, -53.363, -54.563, -55.663, -56.663, -57.663, -58.663, -59.663, -60.663, -61.663]


### Question 4.4

Using function 'simulator' from Q3.3 check that the strategy computed in Q4.3 yield the optimal value also computed in Q4.3.

###### Solution 

I defined the policy (dynmaic_strategy(t,x)) so that I can use the simulator function 

The function works in two steps:-
1. It computes the value function $V_{t+1,x}$ ($i.e. ~$ the cost of being at day $(t+1)$ for any possible value of the stock $x$)
2. It uses the cost vector $V_{t+1,x}$ to compute the optimal contorl $u$ for the day $t$ with the speicifc stock $x$

In [68]:
function dynamic_strategy(t,stock)
    
    V = zeros(T+1,D+1)
    x = stock
    Opt_u = 0

# The code as in the previous block, it just terminates at (T-t) instead of T
    for t=1:T-t
        for x=0:D
            V_hat = Inf
            for u=0:U
                z = 0
                for d=1:D
                    y = x + u - d
                    if (y>0 && y<=D)
                        v = (c*u - p*min(x+u,d)) + V[T-t+2,y+1]
                    elseif y <= 0
                        v = (c*u - p*min(x+u,d)) + V[T-t+2,1]
                    else
                        v = Inf
                    end
                    z += v
                end
                if z < V_hat
                    V_hat = z
                end
            end
            V[T-t+1,x+1] = V[T-t+1,x+1]+P*V_hat
        end
    end
    
# Now, we have the cost vector V[t+1,x] for any x, we will use it to compute 
#         the optimal control for the day t with the specific sctock x
    V_hat1 = Inf
    for u=0:U
        z=0
        for d=1:D
            y = x + u - d
            if (y>0 && y<=D)
                v = (c*u - p*min(x+u,d)) + V[t+1,y+1]
            elseif y <= 0
                v = (c*u - p*min(x+u,d)) + V[t+1,1]
            else
                v = Inf
            end
            z += v
        end
        if z < V_hat1
            V_hat1 = z
            Opt_u = u          # Determining the optimal control 
        end
    end
            
    return Opt_u        # returning the optimal control 
end

dynamic_strategy (generic function with 1 method)

In [69]:
# print(dynamic_strategy(10,5))
simulator(scenarios_mc,dynamic_strategy)

-50.42

Indeed, the simulator yielded the same optimal value $V_{1,1}$ as in $Q4.3$ 

Tested JuMP 0.20 / Julia 1.2